### PR TITLE
Find by url

### PR DESCRIPTION
**Pull request by philipqnguyen**
`Video#find_by_url!(url)` will fetch the video based on a URL that was passed. It first parses the url to find the video_id, then it passes the video_id to `Video#find(video_id)`.

Supported url formats:
`https://www.facebook.com/{account_name}/videos/vb.{alt_account_id}/{video_id}/`
`https://www.facebook.com/{account_name}/videos/{video_id}/`
`https://www.facebook.com/{account_id}/videos/{video_id}/`
`https://www.facebook.com/video.php?v={video_id}`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -94,6 +94,24 @@ If a non-existing video ID is passed into #find, Funky::ContentNotFound will be
...
+video.like_count    # => 1169
+video.comment_count # => 65
+video.share_count   # => 348
+video.view_count    # => 10121
+```
+
+If a non-existing video url is passed into #find_by, Funky::ContentNotFound will be raised.
+
+```ruby
+Funky::Video.find_by(url: 'https://www.facebook.com/video.php?v=doesnotexist')
```
Comment (claudiofullscreen) [inline README.md]: Just as a thought, in ActiveRecord `User.find_by(id: -234)` returns `nil`, does not raise an error.

File: spec/videos/video_spec.rb [hunk 2]
```diff
@@ -87,4 +87,77 @@
       it { expect { video }.to raise_error(Funky::ConnectionError) }
     end
   end
+
+  describe '.find_by(url: url)' do
+    let(:video) {Funky::Video.find_by(url: url)}
+
+    context "given an existing video with this url format: 'https://www.facebook.com/{account_name}/videos/vb.{alt_account_id}/{video_id}/'" do
```
Comment (claudiofullscreen) [inline spec/videos/video_spec.rb]: Can you look at RSpec's `shared_context` to avoid this repetition?

File: lib/funky/url/parser.rb [hunk 3]
```diff
@@ -0,0 +1,20 @@
+module Funky
+  # @api private
+  module URL
+    class Parser
+      attr_reader :url
+      VIDEO_ID_REGEXES = [/videos\/(\d*)/i, /v=(\d*)/i, /vb.\d*\/(\d*)/i]
```
Comment (claudiofullscreen) [inline lib/funky/url/parser.rb]: Should we check that `facebook.com` is present before this? I think http / https / www. are optional but we don't want to accept `youtube.com/videos/1234` as a valid Facebook video URL

File: spec/videos/video_spec.rb [hunk 4]
```diff
@@ -87,4 +87,77 @@
...
+      let :url do
+        'https://www.facebook.com/FOXDanmark/videos/vb.129316923794451/965037490222386/'
+      end
+
+      it { expect(video.id).to be_a(String) }
+      it { expect(video.like_count).to be_an(Integer) }
```
Comment (claudiofullscreen) [inline spec/videos/video_spec.rb]: In all these tests, we don't really care about `like count` being an Integer and so on.

We care about `id` being `965037490222386` (and not, say `129316923794451`).

This PR is just about a "string parser" that knows nothing else about Facebook data.

I would even have all these tests in a different spec file

**Comment by philipqnguyen** [comment]
I updated this branch with your inputs.

**Comment by philipqnguyen** [comment]
I updated the PR to extend one of the REGEX to support additional url variants. These should all work:
"facebook.com/PAGE_NAME/videos/vb.PAGE_ID/#{video_id}"
"facebook.com/PAGE_NAME/videos/vl.PAGE_ID/#{video_id}"
"facebook.com/PAGE_NAME/videos/vl.PAGE_ID/#{video_id}/?type=1"

**Pull request merged**